### DISCUSSION TITLE
New curl library for Python - mcurl

### DISCUSSION THREAD

**Discussion by genotrance** [curl/curl#12956] [Show and tell]
It has been [two years](https://github.com/genotrance/px/discussions/147) since I migrated [Px](https://github.com/genotrance/px) proxy to use libcurl, leveraging all the fantastic work done here. I initially tried to use [pycurl](http://pycurl.io) but it did not allow enough access to the underlying APIs to leverage libcurl as a proxy backend. Moreover, there were no prebuilt binaries posted since 2018 so I put together a lightweight module that exposes the easy and multi interfaces over a ctypes wrapper. It leveraged the system installed version of libcurl (except Windows which ships with a prebuilt binary) and was shipped as a part of Px.

I'm happy to announce that [mcurl](https://github.com/genotrance/mcurl/) is now a standalone [project](https://pypi.org/project/pymcurl/) that ships prebuilt packages for multiple OS and architectures with the latest version of libcurl. I look forward to all feedback from this community. 

It will be great if mcurl can be shared on the [bindings](https://curl.se/libcurl/bindings.html) page.

PS. Looks like pycurl **just** posted binaries [yesterday](https://pypi.org/project/pycurl/7.45.3/#files)!

**Comment by gvanem** [comment]
Trying it with `py -3 setup.py build`, I get:
```py
...
- Extracting MozillaCACerts.v2023.8.22.any.tar.gz
- Copying libraries
Traceback (most recent call last):
  File "F:\gv\Python310\lib\site-packages\cffi\cparser.py", line 336, in _parse
    ast = _get_parser().parse(fullcsource)
  File "F:\gv\Python310\lib\site-packages\pycparser\c_parser.py", line 147, in parse
    return self.cparser.parse(
  File "F:\gv\Python310\lib\site-packages\pycparser\ply\yacc.py", line 331, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "F:\gv\Python310\lib\site-packages\pycparser\ply\yacc.py", line 1199, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "F:\gv\Python310\lib\site-packages\pycparser\ply\yacc.py", line 193, in call_errorfunc
    r = errorfunc(token)
  File "F:\gv\Python310\lib\site-packages\pycparser\c_parser.py", line 1931, in p_error
    self._parse_error(
  File "F:\gv\Python310\lib\site-packages\pycparser\plyparser.py", line 67, in _parse_error
    raise ParseError("%s: %s" % (coord, msg))
pycparser.plyparser.ParseError: <cdef source string>:248:21: before: curl_socklen_t
...
```

No idea what this is about.

**Reply by genotrance** [reply]
mcurl is built using mingw32-x64 on Windows and prebuilt binaries are posted on [PyPI](https://pypi.org/project/pymcurl/). Simply run `pip install pymcurl`.

If you want to build it yourself, you can use `bash build.sh`. It includes various flags to inform setuptools of mingw32 and Py_LIMITED_API and uses delvewheel to include the libcurl binaries in the wheel. I don't have MSVC so haven't tested that code path.

**Comment by bagder** [comment]
#12962
